src/stages/stage5/timeline.c

```diff
@@ -334,197 +135,713 @@ static int stage5_magnetto(Enemy *e, int t) {
+			);
+		}
+
+		play_sound("shot1");
+		WAIT(difficulty);
 	}
+}
 
-	FROM_TO(3000, 3200, 100) {
-		create_enemy1c(VIEWPORT_W/2 + VIEWPORT_W/6*(1-2*(_i&1)), 2000, BigFairy, stage5_lightburst2, -1+2*(_i&1) + 2.0*I);
+TASK(loop_swirls, {
+	int num;
+	int direction;
+	cmplx start;
+	cmplx velocity;
+	cmplx turn;
+	cmplx retention;
+}) {
+	for (int i = 0; i < ARGS.num; i++) {
+		INVOKE_TASK(loop_swirl, {
+			.start = VIEWPORT_W * ARGS.direction + ARGS.start * rng_sreal(),
+			.velocity = ARGS.velocity,
+			.turn = ARGS.turn * rng_sreal() + ARGS.start,
+			.retention = ARGS.retention,
+		});
+		WAIT(10);
 	}
+}
+
+TASK(limiter_fairy, {
+	cmplx pos;
+	MoveParams exit;
+}) {
+	Enemy *e = TASK_BIND(espawn_fairy_red(ARGS.pos, ITEMS(.points = 4, .power = 4)));
+	e->move = ARGS.exit;
+
+	int difficulty = difficulty_value(0.25, 0.50, 0.75, 1);
+	for (int x = 0; x < 400; x++) {
+		double base_angle = carg(global.plr.pos - e->pos);
 
-	FROM_TO(3300, 4000, 90-10*global.diff)
-		create_enemy1c(200.0*I+VIEWPORT_W*(_i&1), 1500, Fairy, stage5_superbullet, 3-6*(_i&1));
+		for (int i = 1; i >= -1; i -= 2) {
+			double a = i * 0.2 - 0.1 * difficulty + i * 3.0 / (x + 1);
+			cmplx aim = cdir(base_angle + a);
+
+			PROJECTILE(
+				.proto = pp_rice,
+				.pos = e->pos,
+				.color = RGB(0.5, 0.1, 0.2),
+				.move = move_asymptotic_simple(10 * aim, 2),
+				.flags = PFLAG_NOSPAWNFADE,
+			);
+		}
 
-	AT(3400) {
-		create_enemy2c(VIEWPORT_W/4, 6000, BigFairy, stage5_laserfairy, 2.0*I, 1);
-		create_enemy2c(VIEWPORT_W/4*3, 6000, BigFairy, stage5_laserfairy, 2.0*I, 1);
+		play_sfx("shot1_loop");
+		WAIT(3);
 	}
+}
 
-	FROM_TO(4200, 5000, 20-3*global.diff) {
-		float f = frand();
-		create_enemy3c(
-			VIEWPORT_W/2+300*sin(global.frames)*cos(2*global.frames),
-			400,
-			Swirl,
-			stage5_swirl,
-			2*cexp(I*M_PI*f)+I,
-			60 + 100.0*I,
-			cexp(0.01*I*(1-2*(f<0.5)))
+TASK(limiter_fairies, {
+	int num;
+	cmplx pos1;
+	cmplx pos2;
+	cmplx exit;
+}) {
+	for (int i = 0; i < ARGS.num; i++) {
+		cmplx pos = ARGS.pos1 + ARGS.pos2 * (i % 2);
+		INVOKE_TASK(limiter_fairy,
+			.pos = pos,
+			.exit = move_linear(ARGS.exit)
 		);
+		WAIT(60);
 	}
+}
 
-	FROM_TO(4320, 4400, 60) {
-		double ofs = 32;
-		create_enemy1c(ofs + (VIEWPORT_W-2*ofs)*(_i&1) + VIEWPORT_H*I, 200, Swirl, stage5_miner, -I);
-	}
+TASK(miner_swirl, {
+	cmplx start;
+	cmplx velocity;
+}) {
+	Enemy *e = TASK_BIND(espawn_swirl(ARGS.start, ITEMS(.points = 2)));
 
-	FROM_TO(4260, 5000, 60) {
-		create_enemy1c(VIEWPORT_W*(_i&1)+120*I, 400, Fairy, stage5_greeter, 6 * (1-2*(_i&1)) + I);
-	}
+	e->move = move_linear(ARGS.velocity);
 
-	AT(5000) {
-		create_enemy1c(VIEWPORT_W/2, 2000, BigFairy, stage5_lightburst, 2.0*I);
+	int difficulty = difficulty_value(8, 6, 4, 3);
+
+	for (int x = 0; x < 600; x++) {
+		PROJECTILE(
+			.proto = pp_rice,
+			.pos = e->pos + 20 * cdir(2.0 * M_PI * rng_real()),
+			.color = RGB(0, 0, 255),
+			.move = move_linear(cdir(2.0 * M_PI * rng_real())),
+		);
```

Comment:
Do not put multiple RNG calls in the same statement. The evaluation order is undefined and may break replays across builds (because C is awesome like that). We had the `a*rand*` facility (and its successor, `RNG_ARRAY` and `vrng_*`) to hack around this problem, but it's usually better to just use variables.

`ninja upkeep` can detect problematic RNG usage, although it has false positives.

Note that this is also sometimes neglected in effect-spawning code, because it ultimately has no effect on game state, but that's still not a good practice.